twilio/http/async_http_client.py

```diff
@@ -0,0 +1,110 @@
+import aiohttp
+import logging
+
+from aiohttp import BasicAuth
+from aiohttp_retry import ExponentialRetry, RetryClient
+from twilio.http import HttpClient, AsyncHttpClient
+from twilio.http.request import Request as TwilioRequest
+from twilio.http.response import Response
+
+_logger = logging.getLogger('twilio.async_http_client')
+
+
+class AsyncTwilioHttpClient(HttpClient, AsyncHttpClient):
+    """
+    General purpose asynchronous HTTP Client for interacting with the Twilio API
+    """
+
+    def __init__(self, pool_connections=True, trace_configs=None, timeout=None, logger=_logger, proxy_url=None,
+                 max_retries=None):
+        """
+        Constructor for the AsyncTwilioHttpClient
+
+        :param bool pool_connections: Creates a client session for making requests from.
+        :param trace_configs: Configuration used to trace request lifecycle events. See aiohttp library TraceConfig
+                              documentation for more info.
+        :param float timeout: Timeout for the requests (seconds)
+        :param logger
+        :param str proxy_url: Proxy URL
+        :param int max_retries: Maximum number of retries each request should attempt
+        """
+        super().__init__(logger, True, timeout)
+        self.proxy_url = proxy_url
+        self.trace_configs = trace_configs
+        self.session = aiohttp.ClientSession(trace_configs=self.trace_configs) if pool_connections else None
+        if max_retries is not None:
+            retry_options = ExponentialRetry(attempts=max_retries)
+            self.session = RetryClient(client_session=self.session, retry_options=retry_options)
```

Comment:
What happens if there is no session already? (`TwilioHttpClient` checks for both `self.session` and `max_retries`)

Reply:
If `self.session` is not created or `None`, the aiohttp-retry `RetryClient` will create its own session that is just an instance of `aiohttp.ClientSession`. Although `client_session` is documented as optional, the package doesn't actually document it creates its own session aside from its code.